### PR TITLE
MAINT: test, refactor design of recursive closures

### PR DESCRIPTION
**Pull request by mattip**
In looking at `flake8 --select F811` local variable reuse, I stumbled across the solution in PR #10621 to the problem of recursive closures (#10620). It seems we can avoid assigning the closure to None by using the design pattern in the [last comment](https://github.com/numpy/numpy/pull/10621#issuecomment-366617500) to that PR. 

I also added a test for the use of recursive closures in `loadtxt`, which exposed another place where we still had cyclical references.

### CODE REVIEW

**Comment by eric-wieser** [comment]
I'm not sure this is an improvement for clarity, although good job finding the other cases. Thoughts, @ahaldane?

**Comment by eric-wieser** [comment]
A possible alternative:
```python
class recursive(object):
    def __init__(self, func):
        self.func = func
    def __call__(self, *args, **kwargs):
        return self.func(self, *args, **kwargs)
```
used as
```python
@recursive
def block_recursion(self, arrays, depth=0):
    ...
    self(arrays[i], ...)

block_recursion(arrays)
```

**Comment by eric-wieser** [comment]
Or even simpler:
```
def recursive(f):
    return f.__get__(f)
```

used the same way as above

**Review by eric-wieser — APPROVED** [review]
Looks like an improvement to me. Let's wait for one more opinion, since `@recursive` was my idea and might be an opinionated one.

**Comment by ahaldane** [comment]
I like this much more than the magical `= None`s from before. Merging is good by me!

Perhaps a slight quibble is that I might call it something more specific, like `recursive_closure`, but feel free to ignore me on that.

**Review by ahaldane — APPROVED** [review]

File: numpy/core/_internal.py [hunk 1]
```diff
@@ -808,3 +808,35 @@ def _is_from_ctypes(obj):
...
+        @recursive
+        def stringify(self, arg0, *arg1):
+            if len(arg1) > 0:
+                return self(*arg1)
+            return str(arg0)
+        stringify(*args)
+
+    '''
+    def __init__(self, func):
+        self.func = func
```
Comment (eric-wieser) [inline numpy/core/_internal.py]: nit: this should probably call `functools.update_wrapper`

**Comment by eric-wieser** [comment]
Whatever, there's little point updating the wrapper if its a local variable anyway. Thanks @mattip!

**Pull request merged**